lib/provisioning/Provisioner.js

```diff
@@ -129,24 +143,170 @@ Provisioner.prototype._validate = function(actual, parameterName) {
 };
 
 // Validate parameters for use in linking/unlinking
-Provisioner.prototype._validateAll = function(parameters) {
-    let parameterNames = ['matrix_room_id', 'remote_room_channel', 'remote_room_server'];
+Provisioner.prototype._validateAll = function(parameters, parameterNames) {
+    if (!parameterNames) {
+        parameterNames = [
+            'matrix_room_id',
+            'remote_room_channel',
+            'remote_room_server',
+            'op_nick',
+            'key'
+        ];
+    }
     for (var i = 0; i < parameterNames.length; i++) {
         this._validate(parameters[parameterNames[i]], parameterNames[i]);
     }
 };
 
+Provisioner.prototype._updateBridgingState = Promise.coroutine(
+    function*(roomId, bridger, status, skey) {
+        let intent = this._ircBridge.getAppServiceBridge().getIntent();
+        try {
+            yield intent.client.sendStateEvent(roomId, 'm.room.bridging', {
+                bridger: bridger,
+                status: status // pending, success, failure
+            }, skey);
+        }
+        catch (err) {
+            throw new Error(`Could not update m.room.bridging state in this room ($err.message)`);
+        }
+    }
+);
+
+Provisioner.prototype._authoriseProvisioning = Promise.coroutine(
+    function*(server, bridger, ircChannel, roomId, opNick, key) {
+        let ircDomain = server.domain;
+
+        // (Matrix) Check power level of user
+        let matrixClient = this._ircBridge.getAppServiceBridge().getClientFactory().getClientAs();
+        var lookup = new StateLookup({
+            client : matrixClient,
+            eventTypes: ['m.room.power_levels']
+        });
+
+        yield lookup.trackRoom(roomId);
+
+        let powerState = lookup.getState(
+            roomId,
+            'm.room.power_levels'
+        )[0];
+
+        let actualPower = powerState.content.users[bridger];
+        let requiredPower = powerState.content.events["m.room.power_levels"];
```

Comment:
This value may not always be defined, in which case you should be basing this off the `state_default` which is always defined. See [the spec](http://matrix.org/docs/spec/client_server/r0.2.0.html#m-room-power-levels) for more info.